pkg/streamingpromql/functions/range_vectors.go

```diff
@@ -50,10 +58,271 @@ var PresentOverTime = FunctionOverRangeVector{
+			minSoFar = p.F
+		}
+	}
+
+	for _, p := range tail {
+		if p.F < minSoFar || math.IsNaN(minSoFar) {
+			minSoFar = p.F
+		}
+	}
+
+	return minSoFar, true, nil, nil
+}
+
+var SumOverTime = FunctionOverRangeVector{
+	SeriesMetadataFunc:             DropSeriesName,
+	StepFunc:                       sumOverTime,
+	NeedsSeriesNamesForAnnotations: true,
+}
+
+func sumOverTime(step types.RangeVectorStepData, _ float64, fPoints *types.FPointRingBuffer, hPoints *types.HPointRingBuffer, emitAnnotation EmitAnnotationFunc) (float64, bool, *histogram.FloatHistogram, error) {
+	fHead, fTail := fPoints.UnsafePoints(step.RangeEnd)
+	hHead, hTail := hPoints.UnsafePoints(step.RangeEnd)
+
+	haveFloats := len(fHead) > 0 || len(fTail) > 0
+	haveHistograms := len(hHead) > 0 || len(hTail) > 0
+
+	if !haveFloats && !haveHistograms {
+		return 0, false, nil, nil
+	}
+
+	if haveFloats && haveHistograms {
+		emitAnnotation(annotations.NewMixedFloatsHistogramsWarning)
+		return 0, false, nil, nil
+	}
+
+	if haveFloats {
+		return sumFloats(fHead, fTail), true, nil, nil
+	}
+
+	h, err := sumHistograms(hHead, hTail, emitAnnotation)
+	return 0, false, h, err
+}
+
+func sumFloats(head, tail []promql.FPoint) float64 {
+	sum, c := 0.0, 0.0
+
+	for _, p := range head {
+		sum, c = floats.KahanSumInc(p.F, sum, c)
+	}
+
+	for _, p := range tail {
+		sum, c = floats.KahanSumInc(p.F, sum, c)
+	}
+
+	return sum + c
+}
+
+func sumHistograms(head, tail []promql.HPoint, emitAnnotation EmitAnnotationFunc) (*histogram.FloatHistogram, error) {
+	var sum *histogram.FloatHistogram
+
+	if len(head) > 0 {
+		sum = head[0].H
+		head = head[1:]
+	} else {
+		sum = tail[0].H
+		tail = tail[1:]
+	}
+
+	// We must make a copy of the histogram, as the ring buffer may reuse the FloatHistogram instance on subsequent steps.
+	sum = sum.Copy()
+
+	for _, p := range head {
+		if _, err := sum.Add(p.H); err != nil {
+			err = NativeHistogramErrorToAnnotation(err, emitAnnotation)
+			return nil, err
+		}
+	}
+
+	for _, p := range tail {
+		if _, err := sum.Add(p.H); err != nil {
+			err = NativeHistogramErrorToAnnotation(err, emitAnnotation)
+			return nil, err
+		}
+	}
+
+	return sum, nil
+}
+
+var AvgOverTime = FunctionOverRangeVector{
+	SeriesMetadataFunc:             DropSeriesName,
+	StepFunc:                       avgOverTime,
+	NeedsSeriesNamesForAnnotations: true,
+}
+
+func avgOverTime(step types.RangeVectorStepData, _ float64, fPoints *types.FPointRingBuffer, hPoints *types.HPointRingBuffer, emitAnnotation EmitAnnotationFunc) (float64, bool, *histogram.FloatHistogram, error) {
+	fHead, fTail := fPoints.UnsafePoints(step.RangeEnd)
+	hHead, hTail := hPoints.UnsafePoints(step.RangeEnd)
+
+	haveFloats := len(fHead) > 0 || len(fTail) > 0
+	haveHistograms := len(hHead) > 0 || len(hTail) > 0
+
+	if !haveFloats && !haveHistograms {
+		return 0, false, nil, nil
+	}
+
+	if haveFloats && haveHistograms {
+		emitAnnotation(annotations.NewMixedFloatsHistogramsWarning)
```

Comment:
(nit) possibly not worth it, but we could consider a helper function for dealing with mixed samples.

Reply:
I think this is simple enough to keep in each function implementation for now - if it becomes more complex in the future we can extract it out to its own function.